pkg/providers/snow/validator.go

```diff
@@ -12,17 +12,42 @@ const (
 	snowballMinSoftwareVersion = "102"
 )
 
-type AwsClientValidator struct {
+// Validator includes a client registry that maintains a snow device aws client map,
+// and a local imds service that is used to fetch metadata of the host instance.
+type Validator struct {
+	// clientRegistry maintains a device aws client mapping.
 	clientRegistry ClientRegistry
+
+	// imds is a local imds client built with the default aws config. This imds client can only
+	// interact with the local instance metata service.
+	imds LocalIMDSClient
+}
+
+// ValidatorOpt updates an Validator.
+type ValidatorOpt func(*Validator)
+
+// WithIMDS returns a ValidatorOpt that sets the imds client.
+func WithIMDS(imds LocalIMDSClient) ValidatorOpt {
+	return func(c *Validator) {
+		c.imds = imds
+	}
 }
 
-func NewValidator(clientRegistry ClientRegistry) *AwsClientValidator {
-	return &AwsClientValidator{
+// NewValidator creates a snow validator.
+func NewValidator(clientRegistry ClientRegistry, opts ...ValidatorOpt) *Validator {
```

Comment:
Is the imds client actually optional? `ValidateControlPlaneIP` will panic if that's missing.

Reply:
added nil check + unit test

Reply:
I think we want to make it an explicit requirement by specifying it as a parameter. That will ensure consumers can't accidentally create invalid objects.

Generally, optional arguments should be for optional things only.

Reply:
the `imds` is optional. The validator is being used in both CLI and controller. As I mentioned in the PR description, we do not need an `imds` client for controller validation -- that's why i made it optional.

Reply:
I see, sorry for the confusion.